securedrop/source_app/__init__.py

```diff
@@ -139,57 +132,17 @@ def load_instance_config() -> None:
     @app.before_request
     @ignore_static
     def setup_g() -> Optional[werkzeug.Response]:
-        """Store commonly used values in Flask's special g object"""
-
         if 'expires' in session and datetime.utcnow() >= session['expires']:
-            msg = render_template('session_timeout.html')
-
-            # Show expiration message only if the user was
-            # either in the codename generation flow or logged in
-            show_expiration_message = any([
-                session.get('show_expiration_message'),
-                logged_in(),
-                was_in_generate_flow(),
-            ])
-
-            # clear the session after we render the message so it's localized
-            session.clear()
-
-            # Persist this properety across sessions to distinguish users whose sessions expired
-            # from users who never logged in or generated a codename
-            session['show_expiration_message'] = show_expiration_message
-
-            # Redirect to index with flashed message
-            if session['show_expiration_message']:
-                flash(Markup(msg), "important")
-            return redirect(url_for('main.index'))
+            # If the session as has expired, redirect to index with flashed message
+            # if the user was in the middle of the generate flow
+            if 'codenames' in session:
+                return clear_session_and_redirect_to_logged_out_page(session)
 
         session['expires'] = datetime.utcnow() + \
             timedelta(minutes=getattr(config,
                                       'SESSION_EXPIRATION_MINUTES',
                                       120))
 
-        # ignore_static here because `crypto_util.hash_codename` is scrypt
-        # (very time consuming), and we don't need to waste time running if
-        # we're just serving a static resource that won't need to access
-        # these common values.
-        if logged_in():
-            g.codename = session['codename']
-            g.filesystem_id = app.crypto_util.hash_codename(g.codename)
-            try:
-                g.source = Source.query \
-                            .filter(Source.filesystem_id == g.filesystem_id) \
-                            .filter_by(deleted_at=None) \
-                            .one()
-            except NoResultFound as e:
-                app.logger.error(
-                    "Found no Sources when one was expected: %s" %
-                    (e,))
-                del session['logged_in']
-                del session['codename']
-                return redirect(url_for('main.index'))
-            g.loc = app.storage.path(g.filesystem_id)
```

Comment:
This authentication logic is now handled by the SessionManager, which is called when the `@login_required` decorator is used in the Flask routes. This ensures that all authentication / session management logic happens in a single code location.